### ISSUE TITLE
Kalman filter example yields non PSD covariances

### ISSUE THREAD

**Issue by BenjaminJCox** [patrick-kidger/diffrax#539]
In the Kalman filter example the matrices Q and R encode the covariance of the white noise process associated with the state and observation process respectively. Unless I am missing something in the code (which seems a direct translation of the reference), these matrices should be PSD by virtue of being the covariance matrices of white noise processes.

In the optimisation we learn a Q matrix that is asymmetric, and therefore not PSD. 
In particular, we learn Q= [[-0.44275677  1.3142775 ], [-1.1867669   0.9120258 ]].

The state process learnt does not make sense: although the filter can be run the state process covariance is not well defined. 

An easy way of fixing this is to work with the matrix square root (e.g. store/learn Q^(1/2) and R^(1/2) then construct Q and R when running the filter). 
In this way we can constrain Q and R to be PSD whilst using an unconstrained optimisation scheme.
This should probably also be done with P0 as well for completeness.

This should be a small change (a few lines and renaming variables).

**Comment by johannahaffner** [comment]
I second this, well spotted!

**Comment by patrick-kidger** [comment]
Good catch! Would you be willing and able to send a PR? (I'll be honest that I don't work with Kalman filters that often, which is probably how this mistake snuck through in the first place... ;) So I'd appreciate another set of eyes on the fix.) The example is here:

https://github.com/patrick-kidger/diffrax/blob/main/examples/kalman_filter.ipynb
